source/game/field/obj/ObjectKinoko.cc

```diff
@@ -0,0 +1,164 @@
+
+ObjectKinokoStem::ObjectKinokoStem(const System::MapdataGeoObj &params, KinokoType type)
+    : ObjectCollidable(params) {
+    m_type = type;
+    load();
+}
+
+/// @addr{0x80807e90}
+ObjectKinokoStem::~ObjectKinokoStem() = default;
+
+/// @addr{0x80807e64}
+const char *ObjectKinokoStem::getKclName() const {
+    return m_type == KinokoType::Light ? "kinoko_kuki" : "kinoko_d_kuki";
+}
+
+ObjectKinokoNm::ObjectKinokoNm(const System::MapdataGeoObj &params) : ObjectKCL(params) {
+    m_type = static_cast<KinokoType>(params.setting(0));
+    m_stemObj = new ObjectKinokoStem(params, m_type);
+}
+
+/// @addr{0x80827a9c}
+ObjectKinokoNm::~ObjectKinokoNm() = default;
+
+/// @addr{0x80827a74}
+const char *ObjectKinokoNm::getKclName() const {
+    return m_type == KinokoType::Light ? "kinoko_g" : "kinoko_d_g";
+}
+
+/// @addr{0x8080761c}
+ObjectKinoko::ObjectKinoko(const System::MapdataGeoObj &params)
+    : ObjectKCL(params), m_objPos(m_pos), m_objRot(m_rot) {
+    m_type = static_cast<KinokoType>(params.setting(0));
+
+    _12e = 0;
+    _144 = 6;
+    _130 = 40;
+    _132 = 10;
+    _12c = 1 % 40;
+    _134 = 0.1f;
+    m_period = 6.0f * F_PI / 40.0f;
+    _13c = 0.0008f;
+    _146 = 0;
+
+    m_stemObj = new ObjectKinokoStem(params, m_type);
+}
+
+/// @adrr{0x80807a54}
+ObjectKinoko::~ObjectKinoko() = default;
+
+/// @addr{0x80807d8c}
+const char *ObjectKinoko::getKclName() const {
+    return m_type == KinokoType::Light ? "kinoko_r" : "kinoko_d_r";
+}
+
+/// @addr {0x80807dac}
+u32 ObjectKinoko::loadFlags() const {
+    return 1;
+}
+
+/// @addr{0x8080782c}
+void ObjectKinoko::calc() {
+    if (_12e == 0) {
+        _12c++;
+    }
+    if (_12c == _130) {
+        _12e++;
+    }
+    if (_132 < _12e) {
+        _12e = 0;
+    }
+    if (_130 < _12c) {
+        _12c = 0;
+    }
+
+    _134 = _13c * static_cast<f32>(_130 - _12c);
+    m_flags |= 8; // |= SCALE_DIRTY
+    m_scale.set(_134 * EGG::Mathf::sin(m_period * static_cast<f32>(_12c)) + 1.0f);
+
+    calcOscillation();
+}
+
+/// @addr{0x80807950}
+ObjectKinokoUd::ObjectKinokoUd(const System::MapdataGeoObj &params) : ObjectKinoko(params) {
+    const u16 periodDenom = std::max<u16>(params.setting(2), 2);
+    _14c = 0;
+    _146 = params.setting(3);
+    _150 = params.setting(4);
+    _152 = params.setting(1);
+    m_periodDenom = periodDenom;
+    m_udPeriod = F_TAU / static_cast<f32>(periodDenom);
+}
+
+/// @addr{0x80807e1c}
+ObjectKinokoUd::~ObjectKinokoUd() = default;
+
+/// @addr{0x80807dfc}
+/// @details The base game does check for the light type, however since m_type never gets set
+/// it'll always be 0 which means it always returns "kinoko_r"
+const char *ObjectKinokoUd::getKclName() const {
+    return "kinoko_r";
+}
+
+/// @addr{0x80807a54}
+void ObjectKinokoUd::calcOscillation() {
+    m_flags |= 1; // |= POSITION_DIRTY;
+    m_pos.y = m_objPos.y +
+            static_cast<f32>(_152) * (EGG::Mathf::cos(m_udPeriod * static_cast<f32>(_146)) + 1.0f) *
+                    0.5f;
```

Comment:
```suggestion
    m_pos.y +=
            static_cast<f32>(_152) * (EGG::Mathf::cos(m_udPeriod * static_cast<f32>(_146)) + 1.0f) *
                    0.5f;
```

Reply:
These are not the same thing, ``m_objPos`` never gets altered(base position), but i *guess* this indeed works too, then we can git rid of ``m_objPos`` altogether

Reply:
actually, that's not true, but from my memory removing this line also doesn't affect anything, perhaps some custom tracks make use of it

Reply:
I'm blind, ignore me